Add restrospetive KEP for ProcMount and target at Beta

haircommander opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Bump ProcMount feature to beta 

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/4265 

<!-- other comments or additional information -->
- Other comments: successor to https://github.com/kubernetes/community/pull/1934

mrunalp commented:
/label lead-opted-in

mrunalp commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,501 @@
+unprivileged containers with user namespaces as it will allow more information
+than is necessary to the program running in the container spawned by kubernetes.
+
+The main use case for this option is to run
```
May want to give other examples such as nested podman.

haircommander replied:
I moved some stuff around to make the use cases clearer

npolshakova commented:
> /label lead-opted-in

Hi @mrunalp, can you add this label to the [KEP issue](https://github.com/kubernetes/enhancements/issues/4265) as well for enhancement tracking if this is targeting 1.29? Thank you!

AkihiroSuda commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+
+#### Story 2
+
+As a cluster admin, I may want to [genuinetools/img](https://github.com/genuinetools/img) inside a kubernetes container.
```
genuinetools/img is no longer maintained.
https://github.com/genuinetools/img/issues/348#issuecomment-982473254

haircommander replied:
dropped!

AkihiroSuda commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+
+2018-05-07: k/community update opened
+2018-05-27: k/kubernetes PR merged with support.
+2023-10-02: KEP opened and targeted at Beta
```
What was the story in 2018-2023?

AkihiroSuda commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+## Drawbacks
+
+
+## Alternatives
```
What about `--oci-worker-no-process-sandbox` ?
https://github.com/moby/buildkit/blob/v0.12.2/examples/kubernetes/job.rootless.yaml#L31

AkihiroSuda commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+
+#### Story 1
+
+As a cluster admin, I would like a way to nest containers within containers and get access to proc entries as if the top level container was running
```
AFAICS you are not interested in getting access to those masked procfs entries.
The actual reason you want to unmask them is because the kernel doesn't allow mounting a new procfs without unmasking them, no?

AkihiroSuda commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+#### Story 3
+
+As a kubernetes user, I may want to build containers from within a kubernetes container. 
+See [this article for more information](https://github.com/jessfraz/blog/blob/master/content/post/building-container-images-securely-on-kubernetes.md).
```
Story 3 seems to overlap with Story 2

haircommander replied:
yeah I dropped 2

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+
+It should be noted that this is different that the host /proc. It is still a newly mounted /proc just the container runtimes will not mask the paths.
+
+Since the only use case for this option is to run unprivileged nested containers, this option should only be allowed or used if the user in the container is not `root`.
```
should or must? If must, it will be enforced at API validation?

haircommander replied:
actually I've just checked--only privileged users can have access to this feature. I have left a note on the dangers of allowing users to access this field. I think a future enhancement could be to allow non root users access to an unmasked proc in the baseline policy (especially if they're confined by a user namespace), but more research has to happen for that.

deads2k replied:
so it's actually the exact opposite of what this sentence said this morning?

haircommander replied:
yeah I misread the code

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+Since the only use case for this option is to run unprivileged nested containers, this option should only be allowed or used if the user in the container is not `root`.
+This can be easily enforced with `MustRunAs`.
+
+Since the user inside is still unprivileged, doing things to `/proc` would be off limits regardless, since linux user support already prevents this.
```
is "doing things" reads or just writes?

haircommander replied:
just writes

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+No, though since its a fairly contained feature I would be surprised if there are issues.
```
Please commit to a manual check anyway.  Particularly cases where kubelets are restarted without interrupting existing containers.

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+  - Condition name: 
+  - Other field: 
+- [x] Other (treat as last resort)
+  - Details: successfully create a pod with a container that uses ProcMountType
```
Behavior silently drops fields, so it won't actually fail.  What's actually needed?  Create of procmounttype and then an initContainer that actually checks the filesystem?

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+  - [Optional] Aggregation method:
+  - Components exposing the metric:
+- [x] Other (treat as last resort)
+  - Details: general health of pods that use this feature.
```
name a specific metric to watch for a change before/after.

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,488 @@
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+Potentially a malicious user given access and running with a root container in the host context could mess with the host processes, but PSA has already
```
Fix how?  There's a "should" above that isn't clear about should or must. And the consequences here are not obvious to me.  And if should isn't enforced today in beta and really ought to be must, then additional consideration/work is required.  We could discuss breaking options since its only alpha today.

haircommander replied:
is the feature being restricted to users in a privileged PSA sufficient to mitigate? I feel like it is

haircommander commented:
updated @deads2k thanks for the review--PTAL

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,479 @@
+  - Condition name: 
+  - Other field: 
+- [x] Other (treat as last resort)
+  - Details: Container created with the Unmasked ProcMountType have paths [here](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L193) as writable, not read only.
```
nit: explode list here please.

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,479 @@
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Can check feature gates enabled on kubelet and kube-apiserver.
```
This means its enabled, doesn't mean its in use.  Is it a query of some kind of pod?  Does the kubelet have a metric?

haircommander replied:
updated with a kubectl query!

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,479 @@
+This requires changes to the CRI runtime integrations so that kubelet will add the specific `unmasked` option.
+
+The main use case for unmasking paths in `/proc` are for a user nesting unprivileged containers within a container. However, having an Unmasked ProcMountType
+is a privileged operation, and thus is part of the [privileged PSS policy](https://github.com/kubernetes/kubernetes/pull/103340). Since a user must have be in
```
this PR says baseline

haircommander replied:
yeah  I was a bit confused at first too, but https://github.com/kubernetes/kubernetes/pull/103340/files#diff-88fde5f23fad7b65d37f3783e7980b18e369b6e03f22c0575611628eec7f476dR63 shows it will fail admission if the value is anything other than default or empty, which explicitly disallows Unmasked for baseline and under

deads2k commented:
This one feels like it's being rushed and the lightness of detail makes it a little hard to be confident in the PRR answers.  Please poke me on slack to circle back after the sig is satisfied with accuracy and level of detail.

AkihiroSuda commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,479 @@
+## Drawbacks
+
+
+## Alternatives
```
Has anyone tried to submit a patch to the kernel to allow mounting a new procfs with masks?

haircommander replied:
that's a cool idea! I have no idea if it's been tried or how difficult it would be

mrunalp replied:
cc: @giuseppe

haircommander replied:
added as an alternative

giuseppe replied:
I'm not aware of any effort to allow it.  AFAIK the risk is that allowing a full proc mount will enable a user to access the previously masked files.

This should be safer together with user namespaces, at least it will be possible to allow nested containers in the userns limited environment.

Tagging @brauner in case he has better news/ideas from the kernel side :-)

dgl replied:
The response was no in 2018: https://patchwork.kernel.org/project/linux-fsdevel/patch/20180404115311.725-1-alban@kinvolk.io/ -- I don't think much has changed really.

I do believe something similar to the LXD approach (I mention below) could be an idea with the advantage it can apply to /sys too.

mrunalp commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,491 @@
+#### GA
+
+- Allowing time for feedback
+- Investigate updates to PSS policies in conjunction with user namespaces
```
I would move this up to beta.

haircommander replied:
good idea, updated

SergeyKanzhelev commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,491 @@
+definition.
+
+By default,`procMount` is `default`, aka the same behavior as today and the
+paths are masked.
```
should we have Masked as an explicit option? If this wasn't a part of an API before, making an explicit Masked option would be a better design

SergeyKanzhelev replied:
it will ensure that the user may force masking as a security feature

haircommander replied:
I am less inclined personally--feels like syntactic sugar. Functionally, default is masked, and a use could just as easily enforce default (see PSS today) as masked

SergeyKanzhelev replied:
discussed with @mrunalp - it may be a good idea to change the descriprion of "default" to say it means that certain subpaths are masked. Link to or oci list https://github.com/kubernetes/kubernetes/blob/598076a71caeaf7720b758eedd5cca5a6364ef62/pkg/securitycontext/util.go#L193-L211

If default meaning is "whatever", I wonder if we can even write an e2e test validating any behavior for the "default"

haircommander replied:
+1 to clarifying the definition: would you like me to capture that work here, and how? 

We can test what default means because it's interpreted by the kubelet, so it's all here. A better situation than if we passed the proc mount type to CRI and then it decided what to do with the field. We can test for the masked value of the files in the container's rootfs.

SergeyKanzhelev replied:
yes, let's mention that we will clarify the semantic of this field

mikebrow replied:
see comment ^ IMO default allows the container runtime to configure .. much as we handle other default isolation tooling options..

SergeyKanzhelev commented on keps/sig-node/4265-proc-mount/README.md:
````diff
@@ -0,0 +1,491 @@
+procMount *ProcMountType
+```
+
+This requires changes to the CRI runtime integrations so that kubelet will add the specific `unmasked` option.
````
are we ready to bump to beta if more changed on runtime is needed?

haircommander replied:
I wrote this KEP as if it was written originally in alpha--this has been done with all supported versions of containerd and cri-o

SergeyKanzhelev replied:
if not needed - can you please list the supported versions.

mikebrow replied:
for `unmasked` does kubelet need a way to (outside mounting..) to figure out this option has been implemented for the container?  Today we only return container id in response to create.. As an alternative we could add a response indicating unmasked is supported, or add a runtime feature.. or put the result / impl detail in info for container status...

SergeyKanzhelev commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,491 @@
+#### Beta
+
+- Add e2e tests for the feature
+- Investigate updates to PSS policies in conjunction with user namespaces
```
both criteria feels to be an alpha scope.

haircommander replied:
PSS policies are already in place--just not with user namespaces. Agreed on e2e test, though I see no reason to block beta because they were not covered before

dgl replied:
+1, I feel pushing this to beta is a bit early when the only safe way to use it is with user namespaces, which are still alpha (it is possible to use it with CRI-O annotations, which we have been doing, but also using that safely has some security footguns).

(cc @rata and @saschagrunert; I believe this could fit the PSS work in https://github.com/kubernetes/kubernetes/pull/118760)

deads2k replied:
Security posture and enforcement must be settled for alpha.

SergeyKanzhelev commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,491 @@
+    UnmaskedProcMount ProcMountType = "Unmasked"
+)
+
+procMount *ProcMountType
```
`nil` is default?

Also current documentaiton is saying: "// Note that this field cannot be set when spec.os.name is windows."

Can we update this by specifying what to expect when it is Windows. Which component will blow out and what is expected Runtime reaction on this

SergeyKanzhelev replied:
this is nit for the implementation. Also would be good to add a test for Windows setting this value

haircommander replied:
agreed

haircommander commented:
thanks @SergeyKanzhelev I updated the proposal to add more flavor of what's been already done, added info on when CRI impls got support, and moved around the e2e test requirement.

SergeyKanzhelev commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,500 @@
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[ ] I/we understand the owners of the involved components may require updates to
```
please check the box

SergeyKanzhelev commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,500 @@
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
```
any module you can think of posting here? The only think comes to mind is the file that setting values of security context. It actually has some decent coverage - I rememer more tests were added there recently.

haircommander replied:
yeah good idea, added securitycontext

SergeyKanzhelev commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,500 @@
+
+- test/e2e_node
+
+additional tests should be added to e2e_node suite to test the adherence of the ProcMount field field
```
```suggestion
- additional tests should be added to e2e_node suite to test the adherence of the ProcMount field
  - Test default behavior actually masks /proc paths.
  - Test Unmasked behavior is not masking /proc paths.
  - Test PSS policies (can we even test it as e2e?)
  - Test that Windows pod cannot be scehduled with the value of ProcMount specifies
```

SergeyKanzhelev commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,500 @@
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
```
I'd guess N/A

haircommander commented:
@SergeyKanzhelev @mrunalp PTAL

mikebrow commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+This is to prevent data from being exposed into a container that should not be.
+However, there are certain use-cases where it is necessary to turn this off.
+
+In 1.12, this was introduced as the ProcMountType feature gate, and has languished since. This KEP is
```
```suggestion
In 1.12, this was introduced as the ProcMountType feature gate, and it has languished in alpha ever since. This KEP is
```

mikebrow commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+## Motivation
+
+For end-users who would like to run unprivileged containers using user namespaces _nested inside_ CRI containers, we need an option to have a `ProcMount`.
+That is, we need an option to designate explicitly turn off masking and setting read-only of paths so that we can mount `/proc` in the nested container as an unprivileged user.
```
```suggestion
That is, we need an option to explicitly turn off masking and setting read-only of paths in `/proc` to allow mounting `/proc` in the nested container as an unprivileged user.
```

mikebrow commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+## Proposal
+
+
+Add a new `string` type field named `ProcMountType` will hold the viable
```
```suggestion
Add a new `string` type named `ProcMountType` for choosing from a set of proc mount isolation mode options.  These
```

mikebrow commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+
+Add a new `string` type field named `ProcMountType` will hold the viable
+options for `procMount` to the `SecurityContext`
```
```suggestion
options for `procMount` will be introduced as a new field in the `SecurityContext` (LinuxContainerSecurityContext/WindowsContainerSecurityContext)
```

SergeyKanzhelev reviewed (approved):
/lgtm

mikebrow commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+options for `procMount` to the `SecurityContext`
+definition.
+
+By default,`procMount` is `default`, aka the same behavior as today and the
```
```suggestion
The default for `procMount` (`default`) allows the container runtime to configure it's own defaults, for example as is done today with the result being the 
```

haircommander replied:
the paths are actually passed from the kubelet today. I have adopted some of your wording but kept that impl detail

mikebrow commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+    // DefaultProcMount uses the container runtime default ProcType.  Most 
+    // container runtimes mask certain paths in /proc to avoid accidental security
+    // exposure of special devices or information.
+    DefaultProcMount ProcMountType = "Default"
```
case?

haircommander replied:
this is what is there today

mikebrow commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+    DefaultProcMount ProcMountType = "Default"
+
+    // UnmaskedProcMount bypasses the default masking behavior of the container
+    // runtime and ensures the newly created /proc the container stays in tact with
```
```suggestion
    // runtime and ensures the newly created /proc the container stays intact with
```

haircommander commented:
@deads2k PTAL

mikebrow reviewed (commented):
see comments, cheers!

haircommander commented:
> for unmasked does kubelet need a way to (outside mounting..) to figure out this option has been implemented for the container? Today we only return container id in response to create.. As an alternative we could add a response indicating unmasked is supported, or add a runtime feature.. or put the result / impl detail in info for container status...

currently all supported runtimes support masking what the kubelet passes, so I don't think this response is necessary

haircommander commented:
@mikebrow thanks for the feedback! I have opened https://github.com/kubernetes/enhancements/pull/4279 as a follow up for the grammatical nits

mikebrow commented:
/LGTM

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- N/A
```
Why are these not applicable?

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+containerd has support in 1.6. See https://github.com/containerd/containerd/pull/5070/commits/07f1df4541d6a81c205d194f4f6ea3a6a95c3e29
+
+The main use case for unmasking paths in `/proc` are for a user nesting unprivileged containers within a container. However, having an Unmasked ProcMountType
+is a privileged operation, and thus is part of the [privileged PSS policy](https://github.com/kubernetes/kubernetes/pull/103340). Since a user must have be in
```
```suggestion
is a privileged operation, and thus is part of the [privileged](https://k8s.io/docs/concepts/security/pod-security-standards/#privileged) Pod Security Standard. Since a user must have be in
```
PodSecurity<i>Policies</i> are no longer part of Kubernetes

sftim replied:
(addressed)

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+This requires changes to the CRI runtime integrations so that kubelet will add the specific `unmasked` option.
+This was done after alpha:
+CRI-O has support in v1.25.0 after https://github.com/cri-o/cri-o/pull/6025/commits/4102586132214263c5d0ae93ec257432653ab82b
+containerd has support in 1.6. See https://github.com/containerd/containerd/pull/5070/commits/07f1df4541d6a81c205d194f4f6ea3a6a95c3e29
```
```suggestion
- CRI-O has support in v1.25.0 after https://github.com/cri-o/cri-o/pull/6025/commits/4102586132214263c5d0ae93ec257432653ab82b
- containerd has support in 1.6. See https://github.com/containerd/containerd/pull/5070/commits/07f1df4541d6a81c205d194f4f6ea3a6a95c3e29
```

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+where nil is default, and is interpreted as "Default" ProcMountType.
+
+When the kubelet is presented with a pod that has a ProcMountType as Unmasked, it will edit the default list of
+masked paths it passes down to the [CRI to be empty](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L216) which it does
```
(nit)
```suggestion
masked paths it passes down to the CRI to be [empty](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L216) which it does
```
“CRI to be empty” is an awkward phrase to use as a hyperlink

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+### Goals
+
+- Add the ProcMountType option to the pod API, allowing users to override the proc mount behavior in a container.
```
```suggestion
- Allow users to customize the proc mount behavior for a Linux container.
```

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+## Motivation
+
+For end-users who would like to run unprivileged containers using user namespaces _nested inside_ CRI containers, we need an option to have a `ProcMount`.
```
```suggestion
Some end users would like to run unprivileged containers that are using user namespaces _nested inside_ containers that are defined in a Pod. The outer containers are visible to the kubelet and, through CRI,
to the container runtime. The kubelet is unaware of the nested inner containers.
Kubernetes defaults to masking the proc mount and setting some paths readonly. When end users want
to run nested containers inside an unprivileged Pod, those users want a way to override that default
behavior.
```

sftim reviewed (commented):
Thanks. I recommend further polishing on this KEP before merge.

haircommander commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+## Summary
+
+Currently the most container runtimes work is by masking and setting as read-only certain paths in `/proc`.
+This is to prevent data from being exposed into a container that should not be.
+However, there are certain use-cases where it is necessary to turn this off.
+
+In 1.12, this was introduced as the ProcMountType feature gate, and has languished since. This KEP is
+a successor to (and heavily based on) https://github.com/kubernetes/community/pull/1934/, updated for the modern era.
```
This is meant to be more of a citation than a summary

sftim replied:
For a KEP, we typically put an actual summary here. That's what I'm recommending. You could keep the “In 1.12, this was introduced” short history if the text before it was a summary of the proposal.

(right now, the summary gives context and history but does not **propose** any change)

haircommander commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+- test/e2e_node
+
+- additional tests should be added to e2e_node suite to test the adherence of the ProcMount field
+  - Test default behavior actually masks /proc paths.
+  - Test Unmasked behavior is not masking /proc paths.
+  - Test PSS policies (if possible to test in e2e)
+  - Test that Windows pod cannot be scehduled with the value of ProcMount specifies
```
yeah

dgl commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+## Drawbacks
+
+
+## Alternatives
```
LXD uses an approach where it mounts an additional `/dev/.lxc/proc` and `/dev/.lxc/sys`, the advantage being the kernel sees an unmasked proc filesystem (it's fine it's mounted anywhere, not just the root). It then uses AppArmor to deny user access to the paths -- meaning it can offer this feature (`security.nesting=true` in LXD terms) regardless of what paths are masked.

I believe an approach like this has benefits -- it means that /sys can also be mounted in unprivileged containers (because usually `/sys` is masked by `/sys/fs/cgroup` anyway). I begun writing a KEP targeting 1.30 (I haven't published it yet, was waiting for 1.29 KEP to be over, as I wasn't expecting this to get rushed at the last minute).

haircommander replied:
> (it's fine it's mounted anywhere, not just the root).

this is a weird limitation... thanks for the context, I'll add the alternative, and we can chat about alternatives. it's possible the runtimes could do something fancy like that instead of masking...

dgl commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+A container running as root user on the host and an unmasked `/proc` could be able to write to the host `/proc`, and thus this privileged designation is appropriate.
+
+### User Stories (Optional)
```
There are potentially some other (very) niche user stories, e.g. key rings are namespaced correctly in user namespaces, so it doesn't make sense to limit access to `/proc/keys` (the user story here is running some credential storage tools that use `keyrings(7)` and wanting to debug them).

dgl commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
```
A user turning this on without user namespaces enforced means root within a container can trivially escape. Pod Security Standards obviously enforces, but worth noting.

sftim replied:
I guess you mean that enforcing the _restricted_ Pod security standard prevents that escape?

dgl commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
```
Risk: Placing more trust in the user namespacing mechanism of the kernel. (Arguably the kernel being responsible for making /proc and /sys secure is better anyway, just worth explicitly drawing attention to.)

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+The main use case for unmasking paths in `/proc` are for a user nesting unprivileged containers within a container. However, having an Unmasked ProcMountType
+is a privileged operation, and thus is part of the [privileged PSS policy](https://github.com/kubernetes/kubernetes/pull/103340). Since a user must have be in
+the privileged policy, they are also trusted to choose the correct user ID and run a workload that won't interfere with the host.
```
This statement appears to be incorrect.  It looks like procMount can be set in baseline: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/pod-security-admission/policy/check_procMount.go#L47-L58

Is it incorrectly categorized?

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+is a privileged operation, and thus is part of the [privileged PSS policy](https://github.com/kubernetes/kubernetes/pull/103340). Since a user must have be in
+the privileged policy, they are also trusted to choose the correct user ID and run a workload that won't interfere with the host.
+
+A container running as root user on the host and an unmasked `/proc` could be able to write to the host `/proc`, and thus this privileged designation is appropriate.
```
Pretty sure we allow baseline to do this today.

haircommander replied:
Though baseline doesn't get access to unmasked proc mount, so the only way a user can get access to both is privileged

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+It has been in support by kube-apiserver and kubelet since 1.12, so version skew is no longer a worry.
```
is the kubelet functionality featuregated?  if so, are there safety concerns when used outside of combination with user namespaces that require coordinated rollout?

haircommander replied:
actually it turns out no. just KAS

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+It has been in support by kube-apiserver and kubelet since 1.12, so version skew is no longer a worry.
```
If the kube-apiserver allows a field to be set and the kubelet has the gate unset, what happens?

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes
```
What is is disabled, how do I actually stop pods with this access?  I'm better we have to drain nodes and that needs to be described here.

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Nothing unexpected.
```
when re-enabled, you have to evict pods again because the existing pods won't actually have their intent fulfilled, right?

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+  - Condition name: 
+  - Other field: 
+- [x] Other (treat as last resort)
+  - Details: Container created with the Unmasked ProcMountType have paths [here](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L193) as writable, not read only.
```
if the umasked proc mount fails, will the pod fail to start and produce a good status message?

haircommander replied:
This kind of error condition would be bubbled up from the runtime presumably, so yes

haircommander replied:
hopefully it's good...

deads2k replied:
> This kind of error condition would be bubbled up from the runtime presumably, so yes

How about a CRIO unit test to check?

haircommander commented:
okay @deads2k @mrunalp @SergeyKanzhelev @mikebrow  @dgl @sftim thanks for y'alls reviews. I have tried to react to all and also retarget the KEP at alpha to follow userns

mrunalp commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+
+2018-05-07: k/community update opened
+2018-05-27: k/kubernetes PR merged with support.
+2023-10-02: KEP opened and targeted at Beta
```
Need to change this line.

haircommander replied:
updated!

mrunalp commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+- additional tests should be added to e2e_node suite to test the adherence of the ProcMount field
+  - Test default behavior actually masks /proc paths.
+  - Test Unmasked behavior is not masking /proc paths.
+  - Test PSS integration (if possible to test in e2e)
```
Suggest a s/PSS/PSA/g in the doc.

sftim commented:
How about making the PR title be “Add retrospective KEP for procMount, at alpha”?

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+## Proposal
+
+
+Add a new `string` named `procMount` to the `securitContext` definition for choosing from a set of proc mount isolation mode options.
```
```suggestion
Add a new `string` named `procMount` to the `securitContexty` definition for choosing from a set of proc mount isolation mode options.
```

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+
+- Feature implemented behind a feature flag
+- Add e2e tests for the feature (must be done before beta)
+- Investigate updates to PSA policies in conjunction with user namespaces
```
Do we need to add security warnings to the documentation for this feature? Maybe also the changelog for the first release after this KEP lands?

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Inconsistent feature gate enablement is the only way. All supported versions of CRIs (CRI-O and containerd) have support for this feature.
```
> All supported versions of CRIs (CRI-O and containerd) have support for this feature.

Does that include _all_ implementations of CRI, or only two widely used ones?

soltysh replied:
This is the type of problem I've mentioned above where you should prepare the feature to work when kube-apiserver or kubelet has it off.

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+###### Does this feature depend on any specific services running in the cluster?
+
+- A CRI implementation that supports this feature
+    - All supported versions currently do.
```
Does that include all implementations of CRI?

haircommander replied:
of containerd+ cri-o

deads2k replied:
update the KEP to be specific about which do.

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+###### What are other known failure modes?
+
+Malicious user gaining access to the host `/proc` with a rootful container
+- protected by Pod Security Admission, which requires the namespace is labeled in privileged PSA.
```
Not sure what this means. The default for Pod security admission is “no enforcement”.

sftim commented on keps/sig-node/4265-proc-mount/kep.yaml:
```diff
@@ -0,0 +1,37 @@
+  - "@mrunalp"
+
+replaces:
+  - "https://github.com/kubernetes/community/pull/1934/"
```
(nit)
```suggestion
  - "https://github.com/kubernetes/community/pull/1934"
```

sftim commented on keps/sig-node/4265-proc-mount/kep.yaml:
```diff
@@ -0,0 +1,37 @@
+
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.29"
```
```suggestion
  alpha: "v1.12"
```

sftim reviewed (commented):
Some more feedback. Thanks for the work so far.

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+Currently most container runtimes mask and set as read-only certain paths in `/proc`.
+This is to prevent data from being exposed into a container that should not be.
+However, there are certain use-cases where it is necessary to turn this off.
+
```
```suggestion
This KEP proposes adding a field to the Pod security context to allow bypassing the usual restrictions.

```

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+
+## Summary
+
+Currently most container runtimes mask and set as read-only certain paths in `/proc`.
```
```suggestion
For Linux containers, common container runtimes mask and set as read-only certain paths in `/proc`.
```

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,497 @@
+
+Add a new `string` named `procMount` to the `securitContext` definition for choosing from a set of proc mount isolation mode options.
+
+The default for`procMount` is `default`, which instructs the container runtime to mask the aforementioned paths.
```
```suggestion
The default for`procMount` is `Default`, which instructs the container runtime to mask the aforementioned paths.
```

sftim commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,485 @@
+
+###### What are other known failure modes?
+
+Malicious user, protected by PSS.
```
(addressed)

pacoxu commented:
> Your updated deadline to make any changes to your KEP is 18:00 PST on Monday 9th October 2023.

https://groups.google.com/g/kubernetes-sig-release/c/emPOBgXxu0Q/m/JxEx-ZrcBgAJ?utm_medium=email&utm_source=footer

The exception is approved by v1.29 release team. However, we may miss the deadline again 🤣. Will this still target to v1.29?

haircommander commented:
> The exception is approved by v1.29 release team. However, we may miss the deadline again 🤣. Will this still target to v1.29?

yeah :( we'll aim for 1.30

stlaz commented:
/sig auth
/triage accepted

ibihim commented:
/assign @ibihim

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,502 @@
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, though a reboot or drain will be necessary to limit any containers that originally got access to Unmasked ProcMountType.
```
how will a node reboot change the pod fields that are present?

deads2k replied:
Oh, i think I see.  Could you enumerate the steps more clearly

1. update featuregate on all kube-apiservers
2. restart all kube-apiservers
3. restart all nodes (to read the new pods, which will now have the fields stripped)

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,502 @@
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The API server will need to recreate the pods that had the request scrubbed of the ProcMountType field (so the original intent can be fulfilled).
```
kube-apiserver doesn't create pods.  Which actor could/should do this?  Or perhaps nothing does?

deads2k replied:
Do workload resources also get their procMountType stripped when the featuregate is off?

haircommander replied:
launching a cluster now to double check though

soltysh commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,504 @@
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Add e2e tests for the feature (must be done before beta)
```
If there exists tests, can you link them in the appropriate section above?

haircommander replied:
none yet I will implement them before moving this to beta

soltysh commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,504 @@
+
+The feature gate is only processed by the API server--Kubelet has no awareness of it. API server will scrub the ProcMount field from the request
+if it doesn't support the feature gate. Since all supported Kubelet versions support ProcMountType field, there's no version skew worry.
+API server can have the feature gate toggled without worrying about doing the same for Kubelets.
```
This is a bit confusing, the feature gate applies to both kube-apiserver and kubelet, in which case you should answer the question what happens when you turn the feature on one, and the other has it disabled?

haircommander replied:
the feature gate only applies to KAS, kubelet doesn't do any feature gate checks, it just accepts from KAS and processes it, regardless of whether the feature is enabled in the kubelet or not

soltysh replied:
I'm guessing that's from a time when feature gates were not a thing. In which case I'd put an explicit information how this is currently implemented and update also `kep.yaml` since it's misleadingly pointing to both kube-apiserver and kubelet as having that feature gate.

soltysh commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,504 @@
+
+###### Are there any tests for feature enablement/disablement?
+
+No
```
I'd say those are required for the beta promotion. We need to make sure that turning it off works as expected.

haircommander replied:
I have left a comment describing my position, PTAL

soltysh commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,504 @@
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+They will be, specifically cases where the Kubelet is restarted.
```
This answer should describe the manual testing scenario performed to ensure the path was tested.

soltysh commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,504 @@
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+N/A
```
Don't we have a reasonable SLO for that kubelet_pod_start_sli_duration_seconds metric? Something like a few seconds? At least you can describe that the metric should maintain values from before the upgrade, for example.

soltysh replied:
see https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#kubernetes-slisslos

haircommander replied:
updated!

soltysh commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,504 @@
+  - Details:
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
```
Missing answer here.

soltysh commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,504 @@
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+Negligible
```
Please be more specific wrt to the size, see the template:
```
Describe them, providing:
  - API type(s):
  - Estimated increase in size: (e.g., new annotation of size 32B)
  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
 ```

soltysh reviewed (commented):
#prr-shadow

haircommander commented:
thanks @soltysh! I have updated for your questions with some additional research I did

soltysh commented on keps/sig-node/4265-proc-mount/kep.yaml:
```diff
@@ -0,0 +1,38 @@
+  - name: ProcMountType
+    components:
+      - kube-apiserver
+      - kubelet
```
Since you mentioned in the doc the feature gate doesn't apply to kubelet, I'd suggest removing it from here.

haircommander replied:
good catch, fixed

soltysh reviewed (commented):
#prr-shadow
prr lgtm

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,525 @@
+
+###### Are there any tests for feature enablement/disablement?
+
+There are not. I'm unsure it will be useful, as the toggling of enablement and the functional enablement of old containers are not related.
```
This must be manually tested to ensure that what is documented in https://github.com/kubernetes/enhancements/pull/4266/files#diff-1d1320b7f2cd6bb7e0cb6ebf236babce8b5bb99eaac25bd7fea57d1c17f21b52R348 functions as expected if it must be used in the field.

haircommander replied:
updated to describe the results of a manual test, and describe plans to add a test

deads2k commented on keps/sig-node/4265-proc-mount/README.md:
```diff
@@ -0,0 +1,524 @@
+###### Are there any tests for feature enablement/disablement?
+
+Yes. I have manually tested feature enablement and disablement on kube-apiserver, and verified that pods are not recreated without
+a drain. There will be an e2e test to verify this as well.
```
if this ends up impractical in practice a unit test showing the field stripping will be sufficient along with a manual test of disablement.

mrunalp commented:
/lgtm
